misc/rawcrawl_util.js

```diff
@@ -0,0 +1,217 @@
+
+        // peer properties
+        var p_v = peer.version;
+        var p_pk = normalizePubKey(peer.public_key);
+        try {
+          var p_ipp = normalizeIpp(peer.ip, peer.port);
+        } catch (error) {
+          var p_ipp = undefined;
+        }
+
+        // Fill in rippled
+        var rippled = rippleds[p_pk]
+        if (rippled) {
+          if(!rippled.ipp)
+            rippled.ipp = p_ipp;
+          if(!rippled.version)
+            rippled.version = p_v;
+        } else {
+          rippleds[p_pk] = { ipp: p_ipp, version: p_v };
+        }
+
+      });
+    });
+    return rippleds;
+  },
+
+  /*
+  * @param {Object} raw crawl
+  * @return {Object} { public_key: {ipp: ipp, version: version, in: count, out: count} }
+  * Takes raw crawl and returns a dictionary of unique rippleds keyed by public
+  * key and with the properties ipp, version, in count and out count.
+  */
+  getRippledsC: function(nodes) {
+    var rippleds = this.getRippleds(nodes);
+    var degrees = this.getDegrees(nodes);
+    _.each(Object.keys(degrees), function(pk) {
+      if (rippleds[pk]) {
+        rippleds[pk].in = degrees[pk].in;
+        rippleds[pk].out = degrees[pk].out;
+      }
+    });    
+    return rippleds;
+  },
+
+  /*
+  * @param {Object} raw crawl
+  * @return {Object} { publickey1,publickey2: 1 }
+  * Takes raw crawl and returns an object of unique edges
+  * edges have in the format "publickey1,publickey2" : 1.
+  */
+  getLinks: function(nodes) {
+    var rippleds = this.getRippleds(nodes);
+    // Create ippToPk using rippleds
+    var ippToPk = {};
+    _.each(Object.keys(rippleds), function(pk) {
+      var ipp = rippleds[pk].ipp
+      if (ipp)
+        ippToPk[ipp] = pk;
+    });
+
+    var links = {};
+    _.each(nodes, function(node) {
+
+      // node properties
+      var n_ipp = Object.keys(node)[0];
+      var n_peers = node[n_ipp].overlay.active;
+
+      _.each(n_peers, function(peer) {
+
+        // peer properties
+        var p_pk = normalizePubKey(peer.public_key);
+        var p_type = peer.type;
+        try {
+          var p_ipp = normalizeIpp(peer.ip, peer.port);
+        } catch (error) {
+          var p_ipp = undefined;
+        }
+
+        // Make link
+        if (p_type) {
+          // Get link
+          if (p_type == "in") {
+            a = ippToPk[n_ipp];
+            b = p_pk;
+          }
+          else if (p_type == "out") {
+            a = p_pk;
+            b = ippToPk[n_ipp];
+          }
+          else if (p_type == "peer") {
+            if (peer.ip) {
+              if (peer.ip.split(":").length == 2) {
+                a = ippToPk[n_ipp];
+                b = p_pk;
+              } else {
+                a = p_pk;
+                b = ippToPk[n_ipp];
+              }
+            }
+          } else {
+            // If type is not in/out/peer
+            throw new Error("Peer has unexpected type")
+          }
+
+          if (a !== undefined && b !== undefined) {
```

Comment:
I think there is an issue with the scope of `a` and `b`
Try adding `var a, b` above after `if (p_type)`